### PR TITLE
[FLINK-10061] [table] [kafka] Fix unsupported reconfiguration in KafkaTableSink

### PR DESCRIPTION
**Pull request by tragicjun**
## What is the purpose of the change

When using KafkaTableSink in "Table.writeToSink(), the following exception is thrown:
`java.lang.UnsupportedOperationException: Reconfiguration of this sink is not supported.`

In `Table.writeToSink()` an invocation to "TableSink.configure()" is made to configure parsed fieldNames and fieldTypes from table query. However, `KafkaTableSink.configure()` doesn't support reconfiguration once a fixed table schema has already been configured. This pull request tries to fix the issue: a quick fix is to change the semantics of "KafkaTableSink.configure()".

## Brief change log

  - Allow reconfiguration to KafkaTableSink even when a fixed table schema exists. The semantics of this configuration therefore is to verify if configured fieldTypes matches with the fieldTypes declared by the fixed table schema. A `ValidationException` would be thrown if they do not match.

## Verifying this change

This change is a trivial fix without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by tragicjun** [comment]
Thanks @twalthr for reviewing. Actually I was using `TableEnvironment.connect()` but got used to `writeToSink` style of defining sinks. I think as long as the `writeToSink` remains there, it is hard to prevent legacy code from using it, which would lead to confusion. 

I assumed that `KafkaTableSink` is internal class and change the API semantics causes no confusion (no one would actually invoke KafkaTableSink.configure() apart from internal classes). It is a quick way of resolving the issue. Please let me know if anything I was missing.

File: flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSink.java [hunk 1]
```diff
@@ -166,8 +167,17 @@ public void emitDataStream(DataStream<Row> dataStream) {
 	@Override
 	public KafkaTableSink configure(String[] fieldNames, TypeInformation<?>[] fieldTypes) {
 		if (schema.isPresent()) {
-			// a fixed schema is defined so reconfiguration is not supported
-			throw new UnsupportedOperationException("Reconfiguration of this sink is not supported.");
+			if (getFieldTypes().length != fieldTypes.length || !Arrays.deepEquals(getFieldTypes(), fieldTypes)) {
+				String fixedFieldTypes = Arrays.deepToString(Arrays.stream(getFieldTypes())
+						.map(s -> s.getTypeClass().getSimpleName()).toArray());
```
[the remark this answers is not in the document]
Comment (tragicjun) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSink.java]: The if statement uses `Arrays.deepEquals()` which is a clean way of comparing arrays. The `Arrays.deepToString()` is just used to build string representation of fieldTypes that are reflected in the exception message. Perhaps the code employs too many **Arrays**, please review it a bit closer :)

[the remark this answers is not in the document]
Comment (twalthr) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSink.java]: Sorry, seems I did a too quick look at the code. The code just looked unnecessary complicated to me. A simple `Arrays.toString()` should also do the job. Btw the checking of the field names is missing nevertheless.

[the remark this answers is not in the document]
Comment (tragicjun) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSink.java]: Agreed, I was thinking too much, `Arrays.toString()` should be fine. As for the checking of the field names, I would like to keep consistent with `TableEnvironment.insertInto()` which only checks field types between source table and table sink, considering field names are really irrelevant.

[the remark this answers is not in the document]
Comment (tragicjun) [reply flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/connectors/kafka/KafkaTableSink.java]: @twalthr please review it again, thanks.

**Comment by tragicjun** [comment]
@fhueske @twalthr have you decided that `Table.writeToSink()` would be dropped? If then, this pull request could be closed.

**Comment by fhueske** [comment]
Hi @tragicjun, I'd prefer to remove the method but we should discuss this change on the dev mailing list and see what others think first.

Thanks, Fabian

**Comment by tragicjun** [comment]
As `Table.writeToSink()` will be deprecated, the pull request could be closed.

**Pull request closed without merging**